DeepState on Windows with MinGW

FabrizioSandri opened the pull request:
# Description
This pull request aims to expand the DeepState framework for Windows platforms so that more people can utilize it. This is primarily done to port the [RcppDeepState](https://github.com/FabrizioSandri/RcppDeepState) fuzz testing framework created in R to the Windows platform. 

# Changes 
DeepState requires a lot of POSIX constructs, such as `fork`, `wait`,and `mmap`, which are not supported by Windows. This is the biggest challenge in porting DeepState to Windows. Let's step-by-step assess each change.


### `DeepState_AllocCurrentTestRun`. 
This function allocates some space for the `DeepState_CurrentTestRun` structure using `mmap` into a shared virtual address space. This allows the child process running the test to access the parent's `DeepState_CurrentTestRun` when running with `fork`. 


The Microsoft official documentation explains how to use memory-mapped files instead of the POSIX function `mmap` in Windows by using `CreateFileMapping` and `MapViewOfFile` [[1]](#references). By doing this, the structure `DeepState_CurrentTestRun` will be accessible by both the parent and child processes. 


### `DeepState_ForkAndRunTest` and `DeepState_RunTestWin`.
It is the responsibility of `DeepState_ForkAndRunTest` to determine whether the test should be executed within a new process or directly in the parent process based on the passed flags. This is done by checking if the `--fork` parameter has been passed. 


The `fork` construct is not available on Windows, although the `CreateProcess` function can be used to go around this [[2]](#references). The core idea is to invoke `CreateProcess` while giving the name of the current running module file via `lpCommandLine` along with several parameters. 


The new process is invoked passing the flag `--direct_run`, which is only defined on Windows platforms, so that DeepState would know that the execution had been started via a `CreateProcess` call: when this parameter is given, DeepState recognizes that it is operating inside the new process and thus goes immediately to the execution of the `TEST` function (following the initialization procedure - shared memory allocation in `DeepState_AllocCurrentTestRun`). Along with `--direct_run` the new process is invoked with `--input_which_test`, telling the child process which test to run. 

**NOTE**: the child process has no way of knowing which test to execute, therefore the idea is to pass the extra flag `--input_which_test` along with the test name.


The Windows API provides the function `GetModuleFileName` which can be used to discover the current module's file name [[3]](#references). 
```c++
/* Get the fully qualified path of the current module */
char command[MAX_CMD_LEN]; 
if (!GetModuleFileName(NULL, command, MAX_CMD_LEN)){
  DeepState_LogFormat(DeepState_LogError, "GetModuleFileName failed (%d)", GetLastError());
  return 0;
}
```

In the meanwhile, the main process  waits for the new "child" process thanks to `WaitForSingleObject`[[4]](#references). Then, using `GetExitCodeProcess`, the exit code of the "child" process is collected [[5]](#references). 


### `DeepState_RunSingle`. 
This function has been added to DeepState since DeepState needs to know when it is being called from a CreateProcess and should then proceed directly to the execution of the provided test. The flag `--direct_run` must have been supplied to the program, as shown in the following lines of code, in order for this function to be called on Windows.

```c++
#if defined(_WIN32) || defined(_MSC_VER)
  if (FLAGS_direct_run) {
    DeepState_RunSingle();
    return 0;
  }
#endif
```

In order to discover the test that the parent process requested to execute, the child process can use the parameter supplied to `--input_which_test`. Iterating through all the tests, it compares `test->test_name` and uses `strcmp` to see whether there is a match. After being located, the test is executed by invoking the standard `DeepState_RunTest` function. When this function terminates, an exit code is generated and communicated to the parent process.



### Stats sys/stat.h
Due to the fact that this library is POSIX-specific, Windows cannot access it. To find out whether directories and files are there, several DeepState methods employ the `stat` function. `GetFileAttributes` is a similar method offered by Windows [[6]](#references).

```c++
#if defined(_WIN32) || defined(_MSC_VER)
  DWORD file_attributes = GetFileAttributes(abspath);
  if (file_attributes == INVALID_FILE_ATTRIBUTES){
    DeepState_Abandon("Specified input path does not exist.");
  }
#elif defined(__unix)
  if (stat(abspath, &statbuf) != 0) {
    DeepState_Abandon("Specified input path does not exist.");
  }
#endif
```

# How has this been tested?
MinGW(with MSYS) has been used to test this on Windows with the Built-In Fuzzer. My intention is to bring RcppDeepState to Windows by compiling DeepState using [Rtools](https://cran.r-project.org/bin/windows/Rtools/), a toolchain based on MinGW, for generating R and R packages from source on Windows [[7]](#references).

### Installation
The developer can use the `pacman` package manager to install all required dependencies before beginning the compilation phase.
```sh
pacman -S  mingw-w64-{i686,x86_64,ucrt-x86_64}-python3 mingw-w64-{i686,x86_64,ucrt-x86_64}-python-setuptools mingw-w64-{i686,x86_64,ucrt-x86_64}-gcc mingw-w64-{i686,x86_64,ucrt-x86_64}-cmake mingw-w64-{i686,x86_64,ucrt-x86_64}-libffi
```

The installation procedure remains the same.
```sh
mkdir deepstate/build && cd deepstate/build
cmake -G "Unix Makefiles" ../
make

# optionally (if running the shell as administrator)
make install
```

### Linking and Test
After DeepState had been successfully built, I tested it by manually creating a basic test harness `test.cpp` and compiling it using the `Makefile` shown below:
```make
DEEPSTATE_LIB="C:/Users/user/Downloads/deepstate-master/build"
DEEPSTATE_INCUDE="C:/Users/user/Downloads/deepstate-master/src/include"

CPPFLAGS=-I${DEEPSTATE_INCUDE}
LDFLAGS=-L${DEEPSTATE_LIB} -Wl,-rpath=${DEEPSTATE_LIB}
LDLIBS=-ldeepstate

test : test.o
	g++ -std=c++11 test.o ${CPPFLAGS} ${LDFLAGS} ${LDLIBS} -o test

test.o : test.cpp
	g++ -std=c++11 -c ${CPPFLAGS} test.cpp -o test.o
```


# Todo
* Since `fnmatch` is not available in Windows at this time, all tests are run on Windows regardless of the `--test_filter` parameter. 


# References
[1] [Creating Named Shared Memory](https://learn.microsoft.com/en-us/windows/win32/memory/creating-named-shared-memory)

[2] [Creating processes](https://learn.microsoft.com/en-us/windows/win32/procthread/creating-processes)

[3] [GetModuleFileName function](https://learn.microsoft.com/en-us/windows/win32/api/libloaderapi/nf-libloaderapi-getmodulefilenamea)

[4] [WaitForSingleObject function](https://learn.microsoft.com/en-us/windows/win32/api/synchapi/nf-synchapi-waitforsingleobject)

[5] [GetExitCodeProcess function](https://learn.microsoft.com/en-us/windows/win32/api/processthreadsapi/nf-processthreadsapi-getexitcodeprocess)

[6] [GetFileAttributes function](https://learn.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-getfileattributesa)

[7] [Using Rtools4 on Windows](https://cran.r-project.org/bin/windows/Rtools/rtools40.html)

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/trailofbits/deepstate?pullRequest=428) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/trailofbits/deepstate?pullRequest=428) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/trailofbits/deepstate?pullRequest=428) it.</sub>

pgoodman commented:
@FabrizioSandri Can you update the README to add notes for building on Windows, and any other relevant windows stuff?

pgoodman reviewed (changes requested):
Left some initial comments. One thing this PR makes me wonder is why not have two versions of various functions? E.g. have a platform-specific files, e.g. `DeepState_UNIX.c`, `DeepState_Win32.c`, and a private `DeepState.h` in the `lib` directory that allows the platform-agnostic `DeepState.c` that uses functions in `DeepState.h` that are defined in the platform-specific files?

pgoodman commented:
Also, I should have said this first, but this is amazing work and thank you :-D

FabrizioSandri commented:
@pgoodman thank you for all of the great feedback. I moved the platform-agnostic constructs inside `Platform.h` as per your suggestion. Furthermore, the implementation of the functions for the particular platform are contained in the files `DeepState_UNIX.c` for Linux and `DeepState_Win32.c` for Windows. I also changed the `CMakeLists.txt` to ensure that both Windows and Linux could compile the program. 

I have added all the instructions for building deepstate on Windows to the README.

FabrizioSandri commented on src/include/deepstate/DeepState.h:
```diff
@@ -29,13 +29,18 @@
 #include <time.h>
 #include <unistd.h>
-#include <fnmatch.h>
+#if defined(_WIN32) || defined(_MSC_VER)
```
Following your suggestion I decided to move platform-agnostic controls inside `Platform.h`. Solved with the last few commits.

FabrizioSandri commented on src/include/deepstate/DeepState.h:
```diff
@@ -619,13 +626,12 @@ static void DeepState_InitInputFromFile(const char *path) {
 
-  size_t to_read = stat_buf.st_size;
+  fseek(fp, 0L, SEEK_END);
+  size_t to_read = ftell(fp);
```
Added the cheks for negative return values from `ftell` and `fseek` with commit 7a41dbe.

FabrizioSandri commented on src/include/deepstate/DeepState.h:
```diff
@@ -753,39 +759,100 @@ static int DeepState_RunTestNoFork(struct DeepState_TestInfo *test) {
+    
+    /* Get the fully qualified path of the current module */
+    char command[MAX_CMD_LEN]; 
+    if (!GetModuleFileName(NULL, command, MAX_CMD_LEN)){
```
I can see from the [`GetModuleFileName`](https://learn.microsoft.com/en-us/windows/win32/api/libloaderapi/nf-libloaderapi-getmodulefilenamea) documentation that the function assures the string's null ending:
>  If the length of the path is less than the size that the nSize parameter specifies, the function succeeds and the path is returned as a null-terminated string.
> 
> If the length of the path exceeds the size that the nSize parameter specifies, the function succeeds and the string is truncated to nSize characters including the terminating null character.

FabrizioSandri commented on src/include/deepstate/DeepState.h:
```diff
@@ -753,39 +759,100 @@ static int DeepState_RunTestNoFork(struct DeepState_TestInfo *test) {
+
+    PROCESS_INFORMATION pi;
+    STARTUPINFO si;
+    DWORD exit_code = 0;
```
Good point; by default, the exit code should be interpreted as `DeepState_TestRunPass` which indicates that the test was successful: fixed this with commit . 

However, I don't get the `int *failed` argument's logic. Do you intend to include it in case the test exits with a code other than `0` without necessarily meaning that the test failed?

pgoodman replied:
@FabrizioSandri it wasn't clear to me that failure was being sufficiently distinguished from other return values. E.g. normally a successful exit code on a UNIX process is `0`. If failure needs to be distinguished and can already be distinguished with the existing return values then I am fine without additional arguments.

FabrizioSandri replied:
Yes, a successful run can already be recognized from a failure. DeepState calls `DeepState_RunSingle()` after a new process is launched, which then calls `DeepState_RunTest()`. The latter terminates with `DeepState_TestRunPass` if the tested function terminates successfully regardless of the exit code, i.e. if the test returns correctly without encountering an exception. This means that the second line of the following snippet of code is reached, and the test is deemed successful regardless of the return value of `test->test_func()`. 

```c++
/* Snippet taken from DeepState_RunTest */
test->test_func(); 
exit(DeepState_TestRunPass);
```

On the other hand `DeepState_RunTest()` is guaranteed to exit with an exit code other than `0` in the event that the test fails, and this situation is thus caught in the condition that I inserted in commit 94f6bb1.

FabrizioSandri commented on CMakeLists.txt:
```diff
@@ -118,11 +118,13 @@ if(NOT CMAKE_BUILD_TYPE)
-set(CMAKE_CXX_FLAGS_RELEASE "-O3")
+if (UNIX)
+  set(DEEPSTATE_PLATFORM_LIB src/lib/DeepState_UNIX.c)
+endif (WIN32)
```
d81e30a solves this.

FabrizioSandri commented on src/include/deepstate/Platform.h:
```diff
@@ -0,0 +1,68 @@
+#ifndef SRC_INCLUDE_DEEPSTATE_PLATFORM_H_
```
aefb01b solves this.

FabrizioSandri commented on src/include/deepstate/Platform.h:
```diff
@@ -0,0 +1,68 @@
+struct DeepState_TestInfo;
+
+#if defined(_WIN32) || defined(_MSC_VER)
+  #include <windows.h>
```
I have rewritten the `IS_REGULAR_FILE` macro function as a C function and placed it inside the platform-specific source files `DeepState_UNIX.c` and `DeepState_Win32.c`. This allowed me to move `#include windows.h>` into `DeepState_Win32.c`.

b9fd18e solves this.

FabrizioSandri commented on src/include/deepstate/DeepState.h:
```diff
@@ -607,7 +624,6 @@ extern void DeepState_Warn_srand(unsigned int seed);
 /* Resets the global `DeepState_Input` buffer, then fills it with the
  * data found in the file `path`. */
 static void DeepState_InitInputFromFile(const char *path) {
```
Forward-declaration and platform specific implementation added in commit 14bd95a.

FabrizioSandri commented on src/include/deepstate/DeepState.h:
```diff
@@ -202,6 +198,34 @@ DEEPSTATE_INLINE static int8_t DeepState_MaxChar(int8_t v) {
 }
 
+
+/* Result of a single forked test run.
```
I created a private `lib/DeepState.h` header file with declarations for private functions. However there are some references, such as `DeepState_RunSingleSavedTestDir` inside `include/DeepState.h` that prevents the declaration `extern bool DeepState_IsRegularFile(char *path);` from being moved(the same problem also for other declarations).

FabrizioSandri commented on src/include/deepstate/Platform.h:
```diff
@@ -0,0 +1,68 @@
+  })
+
+  /* Match a regular expression pattern inside a given string 
+   * TODO: implementation for Windows */
```
Added the link to the issue with commit 3dd4b1b.

agroce commented:
> Also, I should have said this first, but this is amazing work and thank you :-D

Seconded!  Hear Hear!
